src/pie_utils/processor/document/candidate_relation_adder.py

```diff
@@ -0,0 +1,208 @@
+from __future__ import annotations
+
+import json
+import logging
+import random
+from collections import defaultdict
+from typing import Any, Dict, List
+
+from pytorch_ie.annotations import BinaryRelation, Span
+from pytorch_ie.utils.span import is_contained_in
+
+from pie_utils.span.slice import distance
+from pie_utils.statistics import WithStatistics
+
+from ..document import DocumentWithEntitiesRelationsAndPartition
+
+logger = logging.getLogger(__name__)
+
+
+class CandidateRelationAdder(WithStatistics):
+    """CandidateRelationAdder adds binary relations to a document based on various parameters. It
+    goes through combinations of available entity pairs as possible candidates for new relations.
+    Entity pairs which are already part of document as a relation are not added again.
+
+    params:
+        label: label for the new relations to be added
+        use_partition: A boolean parameter to enable partition wise relation creation. If this parameter is enabled then
+                    it uses available partitions in the document otherwise use whole document as single partition
```

Comment:
What does it mean to use partition? (entity pairs not in the same partition are discarded)

Reply:
Ah Ok that's what you meant.

Reply:
fixed with [6a5bf0e](https://github.com/ArneBinder/pie-utils/pull/14/commits/6a5bf0e3b0e083b25ff15b1828eb99579208d66f)